src/processors/default/processMempool.ts

```diff
@@ -0,0 +1,64 @@
+
+
+import _ from 'lodash';
+
+import { Table } from '../../db/db';
+import { pendingMempoolMessages } from '../../triggers/messages';
+import { CrdtOpetation, CrdtState, MempoolMessage, PendingMempoolMessage } from '../../types/message';
+import { Clients, Processor } from '../../types/processor';
+
+export const processMempool: (table: Table) => Processor = 
+  (table) => ({
+
+    name: 'processMempool',
+    trigger: pendingMempoolMessages(table),
+    processorFunction: async (messages: PendingMempoolMessage[] , clients: Clients) => {
+      // ignore inserts for now
+      const staleMessages: MempoolMessage[] = messages.filter(message => message.operation === CrdtOpetation.INSERT);
+
+      // messages are ordered by timestamp.
+      //  group by table, column, entity so that we can easily categorize fresh and stale messages
+      const groupedMessagesUpdates = _.groupBy(
+        messages.filter(message => message.operation === CrdtOpetation.UPDATE),
+        message => `${message.operation}.${message.table}.${message.column}.${message.entityId}`
+      );
+      const entityUpdates: { [id: string]: { [column: string]: string } } = {};
+      const crdtUpdates: CrdtState[] = [];
+
+      Object.values(groupedMessagesUpdates).forEach(groupMessage => {
+        // gets the freshest message out of the mempool
+        const freshestUpdate = groupMessage.pop()!;
+
+        // if the freshest message is fresher than the last time from crdtState, we can process it
+        // and update the crdt state.
+        // otherwise we can discard the message
+        if (freshestUpdate?.lastTimestamp === null || freshestUpdate?.timestamp > freshestUpdate?.lastTimestamp){
+          entityUpdates[freshestUpdate.entityId] = {
+            ...entityUpdates[freshestUpdate.entityId],
+            [freshestUpdate.column]: freshestUpdate.value
+          }
+
+          crdtUpdates.push({
+            table: freshestUpdate.table,
+            column: freshestUpdate.column,
+            entityId: freshestUpdate.entityId,
+            lastTimestamp: freshestUpdate.timestamp,
+          })
+        } else {
+          staleMessages.push(freshestUpdate);
+        }
```

Comment:
i guess we could simplify this nested conditional by:
 - instead of popping for the freshestUpdate, just use groupMessage[0]
 - remove this else condition, since the staleMessages.push will then handle pushing in the first message

Reply:
The freshest message is at the last position in the array, not at position 0. I think your suggestion could still work though.

I can conditionally push the message into the updates arrays, and then remove the else and add everything to staleMesages. It doesn't matter if a fresh update is pushed into staleMessages because all the messages need to be deleted anyway - just as long as we process it before we delete it